File: backend/shared/audit/audit.mjs
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';

const s3 = new S3Client({});
const IFC_BUCKET = process.env.IFC_BUCKET || 'builting-ifc';

// Module-level state — hard-reset on every initAudit call so warm Lambda
// invocations don't carry over decisions from the previous render.
let _renderId = null;
let _stage = null;
let _runId = null;
let _decisions = [];
let _validations = [];

/**
 * Call at handler start. Resets all state — required for warm-start safety.
 */
export function initAudit(renderId, stage, runId) {
  _renderId = renderId;
  _stage = stage;
  _runId = runId;
  _decisions = [];
  _validations = [];
}

/**
 * Append a decision entry. Silently no-ops if initAudit was never called.
 *
 * Required fields: pass, element_id, action, reason
 * Optional fields: before, after, params
 */
export function logDecision(entry) {
  if (!_renderId) return;
  _decisions.push({ ts: new Date().toISOString(), stage: _stage, ...entry });
}

/**
 * Append a validation entry.
 *
 * Required fields: validator, element_id, result, expected, actual, severity
 */
export function logValidation(entry) {
  if (!_renderId) return;
  _validations.push({ ts: new Date().toISOString(), stage: _stage, ...entry });
}

/**
 * Write buffered JSONL files to S3. Non-throwing — audit failure must not
 * fail the render. Resets state so a second call is a no-op.
 */
export async function flushAudit() {
  if (!_renderId) return;
  const prefix = `${_renderId}/diagnostics/${_stage}`;
  const writes = [];

  if (_decisions.length > 0) {
    writes.push(s3.send(new PutObjectCommand({
      Bucket: IFC_BUCKET,
      Key: `${prefix}/decisions.${_runId}.jsonl`,
      Body: _decisions.map(d => JSON.stringify(d)).join('\n'),
      ContentType: 'application/x-ndjson',
    })));
  }

  if (_validations.length > 0) {
    writes.push(s3.send(new PutObjectCommand({
      Bucket: IFC_BUCKET,
      Key: `${prefix}/validation.${_runId}.jsonl`,
      Body: _validations.map(v => JSON.stringify(v)).join('\n'),
      ContentType: 'application/x-ndjson',
    })));
  }

  await Promise.all(writes);
  console.log(`[audit:flush] stage=${_stage} decisions=${_decisions.length} validations=${_validations.length}`);

  // Reset so a second flush (e.g. in error path) is a no-op.
  _decisions = [];
  _validations = [];
}

File: backend/shared/audit/audit.py
"""
audit.py — Pipeline audit log writer for the generate Lambda (Python).

init_audit / log_decision / log_validation / flush_audit

Module-level singleton — hard-reset on every init_audit call so warm Lambda
invocations don't carry over decisions from the previous render.

flush_audit is non-throwing. Callers wrap in try/except and log the warning;
audit failure must not fail the render.
"""
import json
import os
from datetime import datetime, timezone

import boto3

_s3 = boto3.client('s3')
_IFC_BUCKET = os.environ.get('IFC_BUCKET', 'builting-ifc')

_render_id = None
_stage = None
_run_id = None
_decisions = []
_validations = []


def init_audit(render_id: str, stage: str, run_id: str) -> None:
    """Call at handler start. Resets all state — required for warm-start safety."""
    global _render_id, _stage, _run_id, _decisions, _validations
    _render_id = render_id
    _stage = stage
    _run_id = run_id
    _decisions = []
    _validations = []


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec='seconds')


def log_decision(entry: dict) -> None:
    """Append a decision entry. No-ops if init_audit was never called."""
    if not _render_id:
        return
    _decisions.append({'ts': _now(), 'stage': _stage, **entry})


def log_validation(entry: dict) -> None:
    """Append a validation entry. No-ops if init_audit was never called."""
    if not _render_id:
        return
    _validations.append({'ts': _now(), 'stage': _stage, **entry})


def flush_audit() -> None:
    """
    Write buffered JSONL to S3 and reset state.
    Non-throwing — callers must wrap in try/except.
    """
    global _decisions, _validations
    if not _render_id:
        return

    prefix = f'{_render_id}/diagnostics/{_stage}'

    if _decisions:
        body = '\n'.join(json.dumps(d) for d in _decisions)
        _s3.put_object(
            Bucket=_IFC_BUCKET,
            Key=f'{prefix}/decisions.{_run_id}.jsonl',
            Body=body.encode('utf-8'),
            ContentType='application/x-ndjson',
        )

    if _validations:
        body = '\n'.join(json.dumps(v) for v in _validations)
        _s3.put_object(
            Bucket=_IFC_BUCKET,
            Key=f'{prefix}/validation.{_run_id}.jsonl',
            Body=body.encode('utf-8'),
            ContentType='application/x-ndjson',
        )

    print(f'[audit:flush] stage={_stage} decisions={len(_decisions)} validations={len(_validations)}')

    # Reset so a second flush is a no-op.
    _decisions = []
    _validations = []
